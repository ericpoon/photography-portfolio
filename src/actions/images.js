import firebase from '../firebase/firebase';

const database = firebase.database();

export const setImages = images => ({
  type: 'SET_IMAGES',
  images,
});

export const startDownloadImages = () => {
  return (dispatch) => {
    return database.ref('/images').once('value')
      .then((snapshot) => {
        const images = [];
        snapshot.forEach((childSnapshot) => {
          images.push({
            id: childSnapshot.key,
            ...childSnapshot.val(),
          });
        });
        dispatch(setImages(images));
      });
  };
};

export const addImage = image => ({
  type: 'ADD_IMAGE',
  image,
});

export const startAddImage = (image) => {
  return (dispatch) => {
    return database.ref('/images').push(image)
      .then((ref) => {
        dispatch(addImage({
          id: ref.key,
          ...image,
        }));
      });
  };
};

export const editImage = (id, updates) => ({
  type: 'EDIT_IMAGE',
  id,
  updates,
});

export const startEditImage = (id, updates) => {
  return (dispatch) => {
    return database.ref(`/images/${id}`).update(updates)
      .then(() => {
        dispatch(editImage(id, updates));
      });
  };
};

export const deleteImage = id => ({
  type: 'DELETE_IMAGE',
  id,
});

export const startDeleteImage = (id) => {
  return (dispatch) => {
    return database.ref(`images/${id}`).remove()
      .then(() => {
        dispatch(deleteImage(id));
      });
  };
};
