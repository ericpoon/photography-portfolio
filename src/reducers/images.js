const initialState = [];

export default (state = initialState, action) => {
  switch (action.type) {
    case 'SET_IMAGES':
      return action.images;
    case 'ADD_IMAGE':
      return [...state, action.image];
    case 'EDIT_IMAGE':
      return state.map((i) => {
        return i.id === action.id ? { ...i, ...action.updates } : i;
      });
    case 'DELETE_IMAGE':
      return state.filter(i => i.id !== action.id);
    default:
      return state;
  }
};
