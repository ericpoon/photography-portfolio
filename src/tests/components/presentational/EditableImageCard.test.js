import React from 'react';
import { shallow } from 'enzyme';
import EditableImageCard from '../../../components/presentational/EditableImageCard';
import images from '../../fixtures/images';

const { title, subtitle, description } = images[0];
let wrapper,
  onImageClick,
  onSaveClick,
  onDeleteClick;

beforeEach(() => {
  onImageClick = jest.fn();
  onSaveClick = jest.fn();
  onDeleteClick = jest.fn();
  wrapper = shallow(
    <EditableImageCard
      title={title}
      subtitle={subtitle}
      description={description}
      onImageClick={onImageClick}
      onSaveClick={onSaveClick}
      onDeleteClick={onDeleteClick}
    />
  );
});

it('should render editable ImageCard correctly', () => {
  expect(wrapper).toMatchSnapshot();
});

it('should handle onImageClick', () => {
  wrapper.find('div>div').at(0).prop('onClick')();
  expect(onImageClick).toHaveBeenCalled();
});

it('should handle delete image', () => {
  wrapper.find('div>button').at(0).simulate('click');
  expect(onDeleteClick).toHaveBeenCalledWith();
});

it('should handle save image', () => {
  wrapper.find('div>button').at(1).simulate('click');
  expect(onSaveClick).toHaveBeenCalledWith({
    title, subtitle, description, file: undefined,
  });
});

it('should handle discard changes', () => {
  wrapper.setState({ title: 'edited title', subtitle: 'edited subtitle', description: 'edited description' });
  wrapper.find('button').at(2).simulate('click');
  expect(wrapper.state()).toMatchObject({ title, subtitle, description });
});
