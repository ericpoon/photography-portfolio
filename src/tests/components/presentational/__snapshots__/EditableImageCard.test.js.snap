// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render editable ImageCard correctly 1`] = `
<div
  className="image-card"
>
  <div
    className="image-card__image"
    onClick={[Function]}
    style={
      Object {
        "backgroundImage": "url()",
      }
    }
  >
    <button
      className="button button--floating-top-left button--danger"
      onClick={[Function]}
    />
  </div>
  <input
    onChange={[Function]}
    type="file"
  />
  <div>
    <input
      className="image-card__title image-card__title--input
              "
      onChange={[Function]}
      value="Kowloon"
    />
    <input
      className="image-card__subtitle image-card__subtitle--input
              "
      onChange={[Function]}
      value="Dec 31, 2017"
    />
    <textarea
      className="image-card__description image-card__description--textarea
              "
      onChange={[Function]}
      value="Putarem quodque tamquam ii ob an deumque. Me assequi sapores ab verarum. Mei qui religionis nam fas devenietur ponderibus. Junctae invenit plausum differt de ipsamet externo is. Mutationum sui negationem exponuntur"
    />
    <div
      className="image-card__actions"
    >
      <button
        className="button"
        onClick={[Function]}
      >
        Save
      </button>
      <button
        className="button"
        onClick={[Function]}
      >
        Discard
      </button>
    </div>
  </div>
</div>
`;
